### PR TITLE
*: fix wrong replace or insert-on-dup behavior on prefixed clustered index

### LINKED ISSUE #23063: unexpected duplicate error when replace into clustered primary key with prefix column with default value
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table test1(c1 varchar(100) not null default 'xyza', c2 int, primary key(c1(3)) clustered);
replace into test1(c2) values(1);
replace into test1(c2) values(1);
```

### 2. What did you expect to see? (Required)

two replace successed

### 3. What did you see instead (Required)

last replace failed with

```
replace into test1(c2) values(1);
ERROR 1062 (23000): Duplicate entry 'xyz' for key 'PRIMARY'
```

### 4. What is your TiDB version? (Required)

```
Release Version: v4.0.0-beta.2-2243-g7daf7436d
Edition: Community
Git Commit Hash: 7daf7436da5de4e5fea3895e892d10d3790e9102
Git Branch: master
UTC Build Time: 2021-03-03 06:18:37
GoVersion: go1.15
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### LINKED ISSUE #23069: unexpected column value after insert on duplicate update on prefix column clustered index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
 create table tbl_3 ( col_15 text(138) , col_16 varchar(37) default 'yGdboyZqIGDQhwRRc' not null , col_17 text(39) not null , col_18 char(58) default 'vBahOai' , col_19 varchar(470) , primary key idx_12 ( col_16(3),col_17(6),col_15(4) ) , key idx_13 ( col_19(2) ) , key idx_14 ( col_18(3),col_15(2) ) , unique key idx_15 ( col_16(4),col_18(6) ) , unique key idx_16 ( col_17(1) ) ) ;
insert into tbl_3 values ( 'XJUDeSZplXx','TfZhIWnJPygn','HlZjQffSh','VDsepqNPkx','xqtMHHOqnLvcxDpL' ) ;
insert into tbl_3 (col_15,col_17,col_19) values ( 'aeMrIjbfCxErg','HTZmtykzIkFMF','' ) on duplicate key update col_18 = values( col_18 );
```

### 2. What did you expect to see? (Required)

`col_17`  should be `HlZjQffSh`

### 3. What did you see instead (Required)

it be `HlZjQf`

### 4. What is your TiDB version? (Required)

```
Release Version: v4.0.0-beta.2-2243-g7daf7436d
Edition: Community
Git Commit Hash: 7daf7436da5de4e5fea3895e892d10d3790e9102
Git Branch: master
UTC Build Time: 2021-03-03 06:18:37
GoVersion: go1.15
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve?

Issue Number: closes #23069,  closes #23063

Problem Summary:

two question caused those two issues:

- check duplicate logic for replace/insert-on-duplicate need use full row column value but use prefixed index value
- on duplicate update for no-touched column need using full row column value but use prefixed index value

### What is changed and how it works?

What's Changed, How it Works:

- refactor and using `TruncateIndexValue` to truncate row value to prefix value to check index constrict
- `DecodeRawRowData` shouldn't use handle value when handle's column is prefixed column

### Related changes

- n/a

### Check List 

Tests 

- Unit test

Side effects

- n/a
### Release note 

- n/a

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/23091)

### CODE REVIEW

File: executor/batch_checker.go [hunk 1]
```diff
@@ -79,30 +81,38 @@ func getKeysNeedCheck(ctx context.Context, sctx sessionctx.Context, t table.Tabl
...
 		if err != nil {
 			return nil, err
 		}
 	}
 	return toBeCheckRows, nil
 }
 
-func getKeysNeedCheckOneRow(ctx sessionctx.Context, t table.Table, row []types.Datum, nUnique int, handleCols []*table.Column, result []toBeCheckedRow) ([]toBeCheckedRow, error) {
+func getKeysNeedCheckOneRow(ctx sessionctx.Context, t table.Table, row []types.Datum, nUnique int, handleCols []*table.Column,
+	idxHandleCols []*model.IndexColumn, result []toBeCheckedRow) ([]toBeCheckedRow, error) {
```
Comment (coocood) [inline executor/batch_checker.go:115]: Can we pass the primary index directly instead of its `IndexColumn` slice?

File: table/tables/tables.go [hunk 2]
```diff
@@ -898,24 +899,31 @@ func DecodeRawRowData(ctx sessionctx.Context, meta *model.TableInfo, h kv.Handle
...
 			}
-			dtBytes := h.EncodedCol(idxOfIdx)
-			_, dt, err := codec.DecodeOne(dtBytes)
-			if err != nil {
-				return nil, nil, err
-			}
-			dt, err = tablecodec.Unflatten(dt, &col.FieldType, ctx.GetSessionVars().Location())
-			if err != nil {
-				return nil, nil, err
+			if notPrefixCol {
```
Comment (wjhuang2016) [inline table/tables/tables.go:913]: How about
useHandleData := if col.IsCommonHandleColumn(meta) && !types.CommonHandleNeedRestoredData(&col.FieldType) && if notPrefixCol

if useHandleData {
...
} else {
...
}

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**